packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -3204,38 +3235,60 @@ namespace Ifpack2 {
 #undef BLOCKTRIDICONTAINER_DETAILS_COMPUTERESIDUAL
 #endif
         }
-#if defined(KOKKOS_ENABLE_CUDA) && defined(IFPACK2_BLOCKTRIDICONTAINER_ENABLE_PROFILE)
-        cudaProfilerStop();
-#endif
+	IFPACK2_BLOCKTRIDICONTAINER_PROFILER_REGION_END;
       }
     }; 
 
+    template<typename MatrixType>
+    void reduceVector(const ConstUnmanaged<typename ImplType<MatrixType>::impl_scalar_type_1d_view> zz,
+                      const typename ImplType<MatrixType>::local_ordinal_type ncols,
+                      /* */ typename ImplType<MatrixType>::magnitude_type *vals) {
+      IFPACK2_BLOCKTRIDICONTAINER_PROFILER_REGION_BEGIN("BlockTriDi::ReduceVector");	
+      using impl_type = ImplType<MatrixType>;
+      using local_ordinal_type = typename impl_type::local_ordinal_type;
+      using impl_scalar_type = typename impl_type::impl_scalar_type;
+      using magnitude_type = typename impl_type::magnitude_type;
+#if 0
+      const auto norm2 = KokkosBlas::nrm1(zz);
+#else
+      impl_scalar_type norm2(0);
+      Kokkos::parallel_reduce
+        ("ReduceMultiVector::Device",
+         Kokkos::RangePolicy<typename impl_type::execution_space>(0,zz.extent(0)),
+         KOKKOS_LAMBDA(const local_ordinal_type &i, impl_scalar_type &update) {
+          update += zz(i);
+        }, norm2);
```

Comment:
@kyungjoo-kim What is the difference between the above if and else part ? Why is this here instead of in KokkosBlas ?

Reply:
Also we should call cuSparse through Kokkos Blas, when available ...

Reply:
When we do not use CUBLAS, it is not different. When CUBLAS is actually used, it makes difference (in this case, cublas is slower). To prevent slowness, I just put this reduction with kokkos. Once we identify the perf problem on the cublas, we can replace it with that.

Reply:
Please file an issue so we don't forget this ..